Исправление некорректной коневертации простых перечислений

dvakor opened the pull request:
к ишузе #918

dvakor committed:
Добавил заполнение представления конвертируемого перечисления

dvakor committed:
Нормальная реализация получения простых перечислений

dvakor committed:
Пропущенный файл

dvakor committed:
Тесты простых перечислений

dmpas commented on src/ScriptEngine/Machine/Contexts/ContextDiscoverer.cs:
```diff
@@ -135,7 +135,8 @@ private static void RegisterSimpleEnum(Type enumType, RuntimeEnvironment environ
 			{
-				environment.InjectGlobalProperty(instance, enumTypeAttribute.Name, true);
+                GlobalsManager.RegisterInstance(enumType, instance);
+                environment.InjectGlobalProperty(instance, enumTypeAttribute.Name, true);
```
табопробельная боль 😢

dvakor replied:
хм, а как фиксить? я щас по новой табы сделал, гит не видит изменений и в VS по форматированию все ок :((

dmpas replied:
у тебя ВС может быть табы на пробелы сразу заменяет?

EvilBeaver replied:
в VSC можно включить отображение whitespace и посмотреть что там. И можно такое же включить в 1С-конфигураторе и банально скопипастить в него.

dvakor replied:
Дошло только, что в этом файлы отступы табами сделаны, а не проблемами, в остальных вроде пробелы)

EvilBeaver commented on src/ScriptEngine/Machine/Contexts/ContextValuesMarshaller.cs:
```diff
@@ -187,10 +187,8 @@ public static IValue ConvertReturnValue(object objParam, Type type)
-                var osValue = (EnumerationValue)constructor.Invoke(new object[] { null, objParam, DataType.Enumeration });
-                return osValue;
+                var enumImpl = GlobalsManager.GetSimpleEnum(type);
+                return enumImpl.GetPropValue(objParam.ToString());
```
Чет я не уловил... в objParam приходит имя элемента перечисления? Как это работает?

dvakor replied:
эээ, кстати ты прав, мне просто повезло, что имя в енуме совпадает с именем в атрибуте

EvilBeaver replied:
Я даже не об этом... Метод ConvertReturnValue работает, когда мы из шарпа хотим вернуть значение перечисления. В этот метод попадает в objParam конкретное значение (а не его имя) и ToString это странно здесь.

Т.е. вырезанная тобой конструкция делала следующее, если не ошибаюсь:

```c#

public SomeNonOscriptEnum MyMethod()
{
     return SomeNonOscriptEnum.TheImportantEnumValue;
}

```

и далее, этот конвертер на лету создавал CLREnumValueWrapper, который можно уже вернуть в виртуальную машину.

Сейчас этот код вырезан и автоматическое оборачивание обычных "шарповых" перечислений сломалось (или почти сломалось)

я могу ошибаться в логике, этот конвертер писал @dmpas, но смысл его был в применении EnumerationType/EnumItem для простой регистрации перечислений.

dvakor replied:
т.е. он работал вообще с любым перечислением, а не с тем, что размечены атрибутами EnumerationType \ EnumItem? как тут например
![image](https://user-images.githubusercontent.com/4038422/65412890-5cdcc900-de1a-11e9-9d53-8942cd9ac97d.png)

EvilBeaver replied:
Ну чиста теоретически с любым, но имелись в виду только вот эти, что ты показал ) Если исправлять, то на "теоретически любые" не надо нацеливаться, нам нужны только помеченные.

dvakor replied:
Тогда, поясню, о том, что выше, objParam (в моем случае это `AllowedSignEnum.Any`), приводя к строке `ToString()` идет как "Any" в свою очередь у `EnumerationContext`, в который оборачивается этот `AllowedSignEnum` при старте движка имеет свойство `"Any"`, т.к. есть такой алиас в атрибуте `EnumItem`, пожтому работает :)

Собственно мой код упадет, если я создам значение перечисление

```
enum MySuperEnum
{
  [EnumItem("Вася", "Vasia")
  Petia
}
```
т.к. `ToString()` отдаст `Petia`, а свойства будут другие

dvakor replied:
@EvilBeaver Я вот задумался, может кто-то отдает щас не помеченные перечисления в ВК например,  (в движке не нашел), у них же падать начнет, мож тогда вернуть старый и проверку сделать, если это левый енум, тогда по старому, а если помеченные то по новому?

EvilBeaver replied:
Нене, если кто-то отдавал непомеченные перечисления, то это недокументировано было. И вряд ли когда-то работало до конца

dvakor replied:
ок, тогда забью

EvilBeaver replied:
Тогда жду фикса и вливаю

dvakor replied:
Блин только щас понял, что я и не залил фикс, закину сегодня

dvakor committed:
Исправление отступов

dvakor committed:
Переделал получение значения элемента перечисления по атрибуту
Добавил тест на проверку приведения значения перечисления, с именем отличным от .NET имени
